mypy_django_plugin/django/context.py

```diff
@@ -256,7 +256,7 @@ def get_expected_types(self, api: TypeChecker, model_cls: type[Model], *, method
 
                     expected_types[field_name] = model_set_type
 
-            elif contenttypes_in_apps and isinstance(field, GenericForeignKey):
+            elif contenttypes_in_apps and isinstance(field, GenericForeignKey):  # type: ignore[possibly-undefined]
```

Comment:
We should not ignore this rule, unless there's a bug in mypy. It might actually be undefined :)

Reply:
Yeah, so this one is a bit of a mess...

```python
        contenttypes_in_apps = self.apps_registry.is_installed("django.contrib.contenttypes")
        if contenttypes_in_apps:
            from django.contrib.contenttypes.fields import GenericForeignKey
        ...
            elif contenttypes_in_apps and isinstance(field, GenericForeignKey):
                ...
        ...
```

It's protected by the check on `contenttypes_in_apps` at runtime, but mypy doesn't realise this and thinks that `GenericForeignKey` might still be undefined event though it's going to be imported when `contenttypes_in_apps` is true...

I'll have a play, but I suspect that we'll just be trading this for `[assignment,misc]`.

Reply:
In this case, I'm not sure that the cure is better than the disease:

```diff
--- mypy_django_plugin/django/context.py
+++ mypy_django_plugin/django/context.py
@@ -199,9 +199,15 @@ class DjangoContext:
         raise ValueError("No primary key defined")
 
     def get_expected_types(self, api: TypeChecker, model_cls: type[Model], *, method: str) -> dict[str, MypyType]:
-        contenttypes_in_apps = self.apps_registry.is_installed("django.contrib.contenttypes")
-        if contenttypes_in_apps:
-            from django.contrib.contenttypes.fields import GenericForeignKey
+        class FakeGenericForeignKey:
+            pass
+
+        _GenericForeignKey: type[GenericForeignKey | FakeGenericForeignKey]
+
+        if self.apps_registry.is_installed("django.contrib.contenttypes"):
+            from django.contrib.contenttypes.fields import GenericForeignKey as _GenericForeignKey
+        else:
+            _GenericForeignKey = FakeGenericForeignKey
 
         expected_types = {}
         # add pk if not abstract=True
@@ -256,7 +262,7 @@ class DjangoContext:
 
                     expected_types[field_name] = model_set_type
 
-            elif contenttypes_in_apps and isinstance(field, GenericForeignKey):  # type: ignore[possibly-undefined]
+            elif isinstance(field, _GenericForeignKey):
                 # it's generic, so cannot set specific model
                 field_name = field.name
                 gfk_info = helpers.lookup_class_typeinfo(api, field.__class__)
```

I haven't pushed up this change, but have fixed the others. Feel free to apply this if you think it's worth it! 😅

Reply:
Could this be:

```suggestion
            elif contenttypes_in_apps:
                from django.contrib.contenttypes.fields import GenericForeignKey
                if isinstance(field, GenericForeignKey):
```

and we remove the if branch above ?

Reply:
What about:

```python
elif contenttypes_in_apps:
    from django.contrib.contenttypes.fields import GenericForeignKey
    if isinstance(field, GenericForeignKey):
        ...
```

?

Reply:
Ha! Perfect timing! 😆

Reply:
D'oh. That's much better.

Reply:
Changed.